provisionerd/runner.go

```diff
@@ -0,0 +1,872 @@
+package provisionerd
+
+import (
+	"archive/tar"
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"path/filepath"
+	"reflect"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/spf13/afero"
+	"golang.org/x/xerrors"
+
+	"cdr.dev/slog"
+
+	"github.com/coder/coder/provisionerd/proto"
+	sdkproto "github.com/coder/coder/provisionersdk/proto"
+)
+
+type jobRunner interface {
+	start()
+	cancel()
+	isDone() <-chan any
+	fail(ctx context.Context, f *proto.FailedJob) error
+	forceStop()
+}
+
+type runner struct {
+	job                 *proto.AcquiredJob
+	sender              messageSender
+	logger              slog.Logger
+	filesystem          afero.Fs
+	workDirectory       string
+	provisioner         sdkproto.DRPCProvisionerClient
+	updateInterval      time.Duration
+	forceCancelInterval time.Duration
+
+	// mutex controls access to all the following variables.
+	mutex *sync.Mutex
+	// used to wait for the failedJob or completedJob to be populated
+	cond *sync.Cond
+	// closed when the job ends.
+	done         chan any
+	failedJob    *proto.FailedJob
+	completedJob *proto.CompletedJob
+	// active as long as we are not canceled
+	gracefulContext context.Context
+	cancelFunc      context.CancelFunc
+	// active as long as we haven't been force stopped
+	forceStopContext context.Context
+	forceStopFunc    context.CancelFunc
+	// setting this false signals that no more messages about this job should be sent.  Usually this means that a
+	// terminal message like FailedJob or CompletedJob has been sent, but if we are force canceled, we may set this
+	// false and not send one.
+	okToSend bool
+}
+
+type messageSender interface {
+	updateJob(ctx context.Context, in *proto.UpdateJobRequest) (*proto.UpdateJobResponse, error)
+	failJob(ctx context.Context, in *proto.FailedJob) error
+	completeJob(ctx context.Context, in *proto.CompletedJob) error
+}
```

Comment:
That interface is larger, giving access to the `Conn` and to the ability to acquire a new job.  Don't want the runner doing that.

Reply:
![image](https://user-images.githubusercontent.com/7122116/176543940-130b5c7f-010c-40e1-82de-c22ed1b45686.png)
This one seems to essentially be the same, just adds `AcquireJob`, which I don't think is a big deal to expose anyways

Reply:
The provisionerd.Server already has an `acquireJob()` method and it doesn't have this signature, nor should it.

Also, `DRPCProvisionerDaemonServer` has these as methods visible outside the package and we don't want that.

Reply:
I'm not sure creating a single interface that's used in a single place improves clarity around that. I feel like a comment explaining why `acquireJob` shouldn't be passed through would do just as well, because a consumer would just add it if they really wanted to call it anyways.

Reply:
So, with the new structure of Runner in a separate package, we definitely need an interface here:  *Runner can't use the *Server type because it would cause an import loop.

I'm pretty strongly against reusing the ` DRPCProvisionerDaemonServer` interface for this purpose because it requires implementing and exposing the `AcquireJob()` method.

"Keep interfaces small" is generally considered a best practice.

More importantly in this context, though, is that we don't want the Runner acquiring jobs.  That's not the contract: Server acquires the job, Runner updates/fails/completes it.

Moreover, the Server is presently designed with the assumption that there is only one acquired job at a time.  If the Server exposes an external method to allow other things to acquire jobs, that assumption could easily be violated.  We should not leave guns like that lying around for our future selves to get shot in the foot.